Initial Gas Dry Deposition Feature Using Zhang et al. (2003) and ACCESS (Saylor 2013).

drnimbusrain opened the pull request:
This is an initial draft PR for including explicit gas dry deposition through the canopy using the approach from [Zhang et al., 2003](https://doi.org/10.5194/acp-3-2067-2003) and based on previous implementation in ACCESS from [Saylor 2013](https://doi.org/10.5194/acp-13-693-2013) and @btang1 previous work to isolate stand-alone codes for canopy-app.    Here the initial implementation only includes one available gas chemical mechanism for 31 gas species including transported from RACM2 (based on ACCESS implementation), but other gas mechanisms could be added.  Preliminary canopy-app NL options for this include:
![image](https://github.com/user-attachments/assets/3dfcc1f4-06d0-4541-9231-2176a387bea2)

Preliminary results looking at ozone dry deposition in our SE U.S. test domain and point profile example are in very reasonable ranges, 0 - 1 cm/s, and agree roughly with a review paper results from [Clifton et al. (2020)](https://doi.org/10.1029/2019RG000670).  

Spatial plots of ozone dry deposition rate for three canopy-app  levels at 0.5 m vertical resolution (level 5 = 2.5 meters, 10 = 5 meters, and 20 = 10 meters).  Timestamp:  2022-07-01-13:00:00.0000 UTC
![image](https://github.com/user-attachments/assets/df87727b-3570-45c3-9032-e3c481aa4028)
![image](https://github.com/user-attachments/assets/40a22528-9635-4b9f-8263-109be1cd762a)
![image](https://github.com/user-attachments/assets/beef56d2-15bb-4b86-8fa8-59198e85e2d0)

Vertical point profile at Lat=34.03 and Lon = 272.11 inside SE U.S. domain above, Timestamp:  2022-07-01-13:00:00.0000 UTC, canopy height ~ 21 meters
![image](https://github.com/user-attachments/assets/41deb4a5-3fb4-4acd-82fd-9ff4f7679559)
Associated LAD point profile:
![image](https://github.com/user-attachments/assets/f4fa7648-be51-46fc-9be6-8b47118c0316)

Still there is more development, testing, and evaluation/assessment of this draft PR before trying to bring in other resistance-based methods as discussed in Issue #60  (likely in future PRs after this initial approach is working as expected).

- [x] Allow PR #147 to go first and then merge here and test global canopy datasets and runs for this new dry deposition feature branch to other unveil bugs/issues.
- [x] Currently dry deposition to ground surface layer, i.e., soil surface (z=0), is set to zero (see profile above).  Will add ACCESS method of calculating ground soil resistance and deposition rate at z=0, https://github.com/btang1/DryDep_MultiLayer_Code/blob/main/Vd_LEB/DryDep.f90#L182.  
- [x] Add output for other RACM2 gas species dry deposition velocities.
- [x] A current approximation includes use stomatal resistance parameters for [deciduous forest](https://github.com/noaa-oar-arl/canopy-app/blob/feature/drydep_zhang/src/canopy_utils_mod.F90#L1293-L1302) from ACCESS applied to all veg types in canopy-app.  Decide to keep this as a note for now, and will update this later in future PR.  
- [x] Add urban and other vtype including snow/ice, barren, and water resistance/drydep rate for gas species at grids/points outside of contiguous vegetative canopies so they are not masked out by dominant vtypes (@rdsaylor-noaa ).   
- [x] Add aerodynamic resistance at all vtypes (e.g., based on vertically resolved canopy eddy diffusivity inside canopies and  MO theory outside of vertically resolved canopies).


This PR works on closing ideas put forth in Issue #60.

drnimbusrain committed:
Initial commit of feature dry deposition branch (ozone).

drnimbusrain committed:
Updated for working o3_ddep, single point only (need implementation).

drnimbusrain committed:
Adding new ifcanddepgas option.

drnimbusrain committed:
Initial netcdf version of ifcanddepgas for ozone.

drnimbusrain committed:
Merge branch 'develop' into feature/drydep_zhang

drnimbusrain committed:
Merge branch 'develop' into feature/drydep_zhang

drnimbusrain committed:
Added ddep_o3 ncf and txt outputs.

drnimbusrain committed:
Fixed negative temp correction error.

drnimbusrain committed:
Fixed typo...working ddep_o3 netcdf file output.

drnimbusrain committed:
Added ddepgasspec option to choose species.

drnimbusrain committed:
Added chemmechgas_opt and chemmechgas_tot for flexibility.

drnimbusrain committed:
Added all RACM2 ddep gas species to netcdf and text outputs.

drnimbusrain committed:
Changed NL to ddepspecgas_opt = 0

drnimbusrain committed:
Updated README for inital gas dry deposition option.

drnimbusrain committed:
Updated README

drnimbusrain committed:
Updated README.

drnimbusrain committed:
Initial commit of soil ground resistance for gas drydep.

drnimbusrain committed:
Added working gas ddep to soil, only for ozone NetCDF first.

drnimbusrain committed:
Updated NL.

drnimbusrain committed:
Updating NL.

drnimbusrain committed:
Updating NL.

drnimbusrain committed:
Merge branch 'develop' into feature/drydep_zhang

rdsaylor reviewed (commented):
Sanity tests for predicted deposition velocities for species other than ozone should also be performed and documented.  For the ozone deposition velocity results, large cities have no deposition values shown.  This is incorrect and should be rectified.  For example, a city such as Atlanta actually has an extensive vegetative canopy.  Possibly, a separate urban "canopy" parameterization ultimately needs to be developed, but in the meantime a default "urban" deposition velocity should be provided for each species.

drnimbusrain commented:
> Sanity tests for predicted deposition velocities for species other than ozone should also be performed and documented. For the ozone deposition velocity results, large cities have no deposition values shown. This is incorrect and should be rectified. For example, a city such as Atlanta actually has an extensive vegetative canopy. Possibly, a separate urban "canopy" parameterization ultimately needs to be developed, but in the meantime a default "urban" deposition velocity should be provided for each species.

Thanks @rdsaylor-noaa for the great comments!  Working now on soil part and will update my plots here also for other species soon for discussion.

Yes, agreed cities with extensive vegetation fraction need to be included in the future.  This may be a longer target (outside of this PR), however, until we can use land use fractions instead of dominant land use currently (see Issue #122 ), or running canopy-app at higher resolution to capture vegetation surrounding cities (e.g., 3 km or 1 km even).  Also agree though in the meantime we can try and add a default "urban" deposition velocity for each species.  Do you have a recommendation on these urban values to follow for the RACM2 species in your ACCESS?  

However, this also requires some consideration at this point (at least before we get vegfrac or higher resolution working) because the entire premise of the gridded part of canopy-app is to only handle vegetated grid cells (with defined vegetation and based on satellite data contiguous canopy thresholds, e.g., lai, canopy height, and canopy fraction).  This is a must for all calculations in canopy-app to fundamentally treat the vegetative canopy plant distributions.   For now though,  I can see adding an "urban_opt" somewhat similar to how we can treat crop and shrub/scrub/grass options (e.g., "crop_opt" and "ssg_opt") in canopy-app and set a default urban ddep gas for these dominant vegtype category of urban.  

Thanks again!

drnimbusrain committed:
Updated gas ddep to soil for all species using soil temperature.

drnimbusrain committed:
Default NL fix.

drnimbusrain commented:
Now adding gas soil resistance and dry deposition rates for under canopy conditions, they are no longer zero at z = 0.  Here is ozone spatial at Level = 0:

Same date/time as above, Timestamp: 2022-07-01-13:00:00.0000 UTC
![image](https://github.com/user-attachments/assets/b3e9ae78-70cd-497f-bc0f-8685d344022f)

And the corresponding vertical point profile with non-zero ddep to soil under relevant canopies:
![image](https://github.com/user-attachments/assets/01d6735a-e461-4d2d-b420-de71f9f7a62f)

drnimbusrain committed:
Updated point text files to be conistent location as before.

drnimbusrain committed:
Updating point txt files again for right soil temps at locations/times.

drnimbusrain committed:
Merge branch 'develop' into feature/drydep_zhang

drnimbusrain commented:
@rdsaylor-noaa Thinking about your comment more on urban vtype deposition, then this would also apply to other vtype categories that canopy-app drydep does not fundamentally treat (e.g., snow/ice, barren, or water, particularly in global applications). I think this then begins to deviate away from what canopy-app was developed to do, i.e., representative a vegetative canopy-only modulating process model aimed at developing relevant parameterizations that interact with host UFS model components (e.g., LSM) to impact such relevant processes (e.g., in-canopy winds, photolysis attenuation, diffusion, biogenic emissions, drydep etc.).  We can think about adding these vtypes for deposition in canopy-app stand-alone (using a vtype "other_opt" or "not_canopy_opt" category with simplifications/assumptions), but would envision the host model to more appropriately handle such other vtype categories rather than what canopy-app interacts with thus modulating relevant canopy processes.

drnimbusrain committed:
Merge branch 'develop' into feature/drydep_zhang

drnimbusrain committed:
Global test OK, updated tka profile with temp at hyblev1 inputs, README, and slurm script.

drnimbusrain committed:
Reverted NL to default.

drnimbusrain commented:
Global tests now work OK [after some adjustments](https://github.com/noaa-oar-arl/canopy-app/pull/149/commits/6c9b9a1a8b31ff473954813b42367cb7e3e6e9a4), and initially needing to run lower vertical canopy model resolution due to memory/time constraints.  This satisfies above task 1.

Shown here for ozone dry deposition at soil level 0 (z = 0) and canopy level 1 (z = 10 m) in this case (5 model levels at 10 m vertical resolution), only for forested canopy regions represented by GEDI canopy heights above thresholds and valid vtypes:
![image](https://github.com/user-attachments/assets/2e1db51b-c5c1-4942-8560-265a505e2a9a)

![image](https://github.com/user-attachments/assets/5827fe5b-9e18-4878-a4b5-5d15a5aee47e)

canopy-app is highly tunable in the NL and can extend canopy coverage in areas outside of tall forest canopies using lowest canopy thresholds (i.e.,  `lai_thresh  =  0.0`, `cf_thresh   =  0.0`, and   `ch_thresh   =  0.0`)  and turning on `ssg_opt` and `crop_opt` to get dry deposition for more low-lying vegetation globally.  Here turning on only dry deposition option  in the NL and selecting only ozone species output we can use more levels (100 levels at 0.5 m resolution), and output in-canopy for different levels to show larger coverage:

Level 0 (z=0 m)
![image](https://github.com/user-attachments/assets/0f77c3c2-f030-4592-a046-07e6bdbb78e9)
Level 1 (z = 0.5 m)
![image](https://github.com/user-attachments/assets/a288ccc3-b076-4b0e-a075-5782f6e19a98)
Level 2 (z = 1.0 m)
![image](https://github.com/user-attachments/assets/3dd98e27-4be4-4dad-bf00-80747a4e82ab)



However, as discussed above, canopy-app still only handles vegetative canopies and does not currently include urban, snow/ice, barren, or water vtype categories.  

I will note compared to [Clifton et al. (2020) review paper](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2019RG000670) with big-leaf ozone dry deposition (Fig. 8 latitudinal averages based on an ensemble of global models, mostly using the Wesely, 1989 scheme), the canopy-app values are very reasonable within the peak latitude band of 45-60 N during August (canopy-app plot shown for July):
![image](https://github.com/user-attachments/assets/c18bec9b-e907-4699-a0ee-19e33e97bbf8)

drnimbusrain committed:
Initiating speices other than o3 to FillValues

drnimbusrain committed:
Formatting.

drnimbusrain committed:
Changed "rate" to "velocity" in output netCDF files.

drnimbusrain commented:
**Mostly all other ddep species look reasonable, e.g., NO2 and HNO3 below:**

![image](https://github.com/user-attachments/assets/d89c8bc6-3fe9-48e0-8f74-de5573731422)   
![image](https://github.com/user-attachments/assets/e04b0d62-5ce6-43f1-8b3f-ef50606eada7)
![image](https://github.com/user-attachments/assets/82772c9f-1e32-4c6f-8b5c-4b5fdbdfdac2)

![image](https://github.com/user-attachments/assets/bf337931-0b72-4b9f-b589-7809523406bc)
![image](https://github.com/user-attachments/assets/602758be-18fe-43a9-b8ed-46a7b65906d5)
![image](https://github.com/user-attachments/assets/3d5b3e6e-6146-463b-b87a-43fb9ae3e1cc)

**However, three species have very low, constant in-canopy deposition velocities, e.g., NO (values ~ 6.352326e-07 cm/s) and CO (3.276465e-07 cm/s) and CH4 (~ 4.680663e-07 cm/s) throughout the canopy levels (at z=0, soil ddep velocities look OK):**
![image](https://github.com/user-attachments/assets/d41323da-b733-41bc-b380-19a41eccb082)
![image](https://github.com/user-attachments/assets/83790b0c-5bb3-40b3-a710-f138ceb5277f)
![image](https://github.com/user-attachments/assets/6a8cc037-95ed-443b-9afe-180d48524371)

![image](https://github.com/user-attachments/assets/16dcfc63-1bc3-48a3-ad1f-bb5c749562ff)
![image](https://github.com/user-attachments/assets/8db6243c-c0c7-4c80-8cef-44ff4ca2c93f)
![image](https://github.com/user-attachments/assets/75e06ba9-15af-416b-ab53-3f329e033494)

![image](https://github.com/user-attachments/assets/e2f1fe2f-dd49-4361-90f3-bafb1a4fdcc9)
![image](https://github.com/user-attachments/assets/a3aa6cbc-4e67-49da-abda-ea157951b7e7)
![image](https://github.com/user-attachments/assets/72657e7b-d08c-4b31-a1d0-0b7e2bdb2c3c)


@rdsaylor-noaa I do notice that for these species  NO, CO, CH4,  the [Henry's Law coefficients are very low, at least an order of magnitude smaller than all other species](https://github.com/noaa-oar-arl/canopy-app/blob/feature/drydep_zhang/src/canopy_utils_mod.F90#L1586C26-L1589) at 1.9D-03, 9.8D-04, and 1.4D-03, respectively.  Likely this is the reason for their much lower in-canopy ddep velocities.  Same with NO3,   O3P,    and O1D, with orders of magnitude smaller Henry's law coefficients.

rdsaylor commented:
@drnimbusrain I think this is a general philosophical difference that we have about what canopy-app should be.  I understand your reluctance to expand the scope of the app, but what it leads to is a divided treatment of the deposition velocities, with some being calculated in canopy-app and the rest in the "host" model. For water, ice/snow and bare ground this separation is inelegant, but probably not a significant issue.  For urban LU with heavy forest coverage (like Atlanta or Baltimore/Washington) the host model's theoretical framework may or may not be consistent with what is used in canopy-app.

rdsaylor reviewed (approved):
Notwithstanding our differences in philosophy about what canopy-app should be, I approve of what has been done in this PR.

drnimbusrain commented:
> Notwithstanding our differences in philosophy about what canopy-app should be, I approve of what has been done in this PR.

Thanks Rick!  Let me have another look at treating urban, barren, snow/ice, and water drydep velocities for these RACM2 species.  Do you have a quick place I can get the urban, barren, snow/ice, and water surface resistances for these species?  I guess for barren I can treat this the same as I am using for soil surface, but need default urban, snow/ice, and water resistances. I can otherwise go in and look at CMAQ etc.

rdsaylor commented:
Probably best to use whatever CMAQ uses.
--
Rick D. Saylor, PhD
Physical Scientist
Atmospheric Turbulence and Diffusion Division
NOAA Air Resources Laboratory
Oak Ridge, TN 37830


On Fri, Mar 7, 2025 at 2:30 PM Patrick Campbell ***@***.***>
wrote:

> Notwithstanding our differences in philosophy about what canopy-app should
> be, I approve of what has been done in this PR.
>
> Thanks Rick! Let me have another look at treating urban, barren, snow/ice,
> and water drydep velocities for these RACM2 species. Do you have a quick
> place I can get the surface resistances for these species/land-use types? I
> can otherwise go in and look at CMAQ etc.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/noaa-oar-arl/canopy-app/pull/149#issuecomment-2707245890>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGCCI52U7PWHFNNKZEKVHC32THXVPAVCNFSM6AAAAABX5RZQE6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBXGI2DKOBZGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: drnimbusrain]*drnimbusrain* left a comment
> (noaa-oar-arl/canopy-app#149)
> <https://github.com/noaa-oar-arl/canopy-app/pull/149#issuecomment-2707245890>
>
> Notwithstanding our differences in philosophy about what canopy-app should
> be, I approve of what has been done in this PR.
>
> Thanks Rick! Let me have another look at treating urban, barren, snow/ice,
> and water drydep velocities for these RACM2 species. Do you have a quick
> place I can get the surface resistances for these species/land-use types? I
> can otherwise go in and look at CMAQ etc.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/noaa-oar-arl/canopy-app/pull/149#issuecomment-2707245890>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGCCI52U7PWHFNNKZEKVHC32THXVPAVCNFSM6AAAAABX5RZQE6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBXGI2DKOBZGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

drnimbusrain committed:
Modified canopy thresholds added ddep gas for barren vtypes.

drnimbusrain commented:
> Probably best to use whatever CMAQ uses. -- Rick D. Saylor, PhD Physical Scientist Atmospheric Turbulence and Diffusion Division NOAA Air Resources Laboratory Oak Ridge, TN 37830
> […](#)
> On Fri, Mar 7, 2025 at 2:30 PM Patrick Campbell ***@***.***> wrote: Notwithstanding our differences in philosophy about what canopy-app should be, I approve of what has been done in this PR. Thanks Rick! Let me have another look at treating urban, barren, snow/ice, and water drydep velocities for these RACM2 species. Do you have a quick place I can get the surface resistances for these species/land-use types? I can otherwise go in and look at CMAQ etc. — Reply to this email directly, view it on GitHub <[#149 (comment)](https://github.com/noaa-oar-arl/canopy-app/pull/149#issuecomment-2707245890)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGCCI52U7PWHFNNKZEKVHC32THXVPAVCNFSM6AAAAABX5RZQE6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBXGI2DKOBZGA> . You are receiving this because you were mentioned.Message ID: ***@***.***> [image: drnimbusrain]*drnimbusrain* left a comment ([noaa-oar-arl/canopy-app#149](https://github.com/noaa-oar-arl/canopy-app/pull/149)) <[#149 (comment)](https://github.com/noaa-oar-arl/canopy-app/pull/149#issuecomment-2707245890)> Notwithstanding our differences in philosophy about what canopy-app should be, I approve of what has been done in this PR. Thanks Rick! Let me have another look at treating urban, barren, snow/ice, and water drydep velocities for these RACM2 species. Do you have a quick place I can get the surface resistances for these species/land-use types? I can otherwise go in and look at CMAQ etc. — Reply to this email directly, view it on GitHub <[#149 (comment)](https://github.com/noaa-oar-arl/canopy-app/pull/149#issuecomment-2707245890)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGCCI52U7PWHFNNKZEKVHC32THXVPAVCNFSM6AAAAABX5RZQE6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBXGI2DKOBZGA> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Thanks Rick!  OK, I have added barren land/vtype easily using the same soil resistance parametrization, and I have updated canopy ssg and crop thresholds a bit so that it is easy to override and use those grid cells as well.  Here is the updated map so far with much greater ddep coverage:

![image](https://github.com/user-attachments/assets/c37c5122-17bd-45fb-a305-c388d1a14f60)
 
I will now move to adding snow/ice, water, and urban resistance functions/subroutines and allow for canopy-app drydep to those vtypes (even outside of vegetative canopy regions).  Will update full global coverage map, hopefully soon.

drnimbusrain committed:
Updating NL for new ssg and crop options.

drnimbusrain committed:
Adding ddep gas snow/ice category and dynamic with new "snowc_ave" input variable.

drnimbusrain commented:
Added new snow/ice vtype category functionality (based off of CMAQ snow resistance and reactivity relative to HNO3) as well as dynamic control on ground (both beneath vegetative canopies and outside of canopies, e.g., barren ground and snow/ice) to account for if covered by diagnostic snow using the GFS "snowc_ave" variable.

![image](https://github.com/user-attachments/assets/478196c3-86d2-46f2-8cd3-fae024340d44)

Also added new snow cover threshold in NL for flexibility:
![image](https://github.com/user-attachments/assets/ca184c88-3b77-486c-ab16-eaed140b2c8c)

drnimbusrain committed:
Updated README.

drnimbusrain committed:
Updated NL to move ssg and crop opts around.

drnimbusrain committed:
Added snow/ice for all species in txt I/O.

drnimbusrain committed:
Added snow and ice conditions over land.

drnimbusrain committed:
Updated README.

drnimbusrain commented:
Added new "icec_set" NL parameter for flexibility, which also will be useful (in conjunction with "snowc_set") for upcoming water points/grids.
![image](https://github.com/user-attachments/assets/3f62d906-d10d-47e5-9379-bd2836f13873)

drnimbusrain committed:
Added urban drydep gas, new can_opts and updated ssg_opt and crop_opt.

drnimbusrain commented:
@rdsaylor-noaa  Added new urban canopy dry deposition scheme (based on Shen and Gao (2018); https://doi.org/10.1016/j.buildenv.2018.02.046), which depends on reaction probability (gamma, namelist setting below) for different building surfaces (default assumed a constant value of 5D-5 for the range in urban surfaces), and in this case the average molecular speed from Maxwell Boltzmann distribution dependent on species.

![image](https://github.com/user-attachments/assets/90cd09f0-48c4-4d1e-9af0-50fbcd21c28c)

Also new can_opt and modified ssg_opt and crop_opt to better  handle missing satellite observations of ch, canfrac, and lai for processes such as drydep (and biogenic emissions too) to extend drydep to all grids/points.Now have drydep over all land points including vegetative canopies and low-lying vegetation (and soils beneath canopies, i.e., soil ), as well as other land vtypes including barren lands/soils, snow/ice, and urban grids/points:

![image](https://github.com/user-attachments/assets/a16a0e15-0265-4aa4-9a21-54cd2d1af8fb)
Note:  This also requires setting the following NL tuning parameters to get all grid points, including any missing satellite ch, cf, and lai, thus overriding with user set constant values for these categories.
```
!Contiguous canopy model thresholds
  lai_thresh  =  0.0
  cf_thresh   =  0.0
  ch_thresh   =  0.0
!Canopy, crop, and shrub/savanna/grass extension options
  can_opt     =  1
  can_laiset  =  4.0
  can_cfset   =  0.5
  can_chset   =  10.0
  crop_opt    =  1
  crop_laiset =  0.2
  crop_cfset  =  0.2
  crop_chset  =  3.0
  ssg_opt     =  1
  ssg_laiset  =  0.2
  ssg_cfset   =  0.2
  ssg_chset   =  1.0
```

Next, I will update final vtype category for drydep to water grids/points.

drnimbusrain committed:
Updated README.

drnimbusrain committed:
Added drydep gas to water for all species.

drnimbusrain commented:
Following CMAQ methodology, I have added drydep gas for all species to dominant water grids/points.   Now there is full coverage across the globe in terms of drydep gas for all vtypes, including all canopy resolved grid cells (z = 1 to n layers), soils/ground beneath canopies (z=0), and all points outside of contiguous canopies set by NL (i.e.,g barren, snow/ice, urban, and water).  

For example, July 01, 2022 1300 UTC ozone drydep with canopy-app reduced vert resolution (modlays = 5 and modres = 10 m):

Level0, z = 0 (surface)
![image](https://github.com/user-attachments/assets/dd28bcb3-0202-47e7-a41d-421771142b01)

Note: drydep to water is order of magnitude smaller than land, i.e., reducing scale for ozone drydep to surface, z=0 (as shown above).  This helps show the contours over water.
![image](https://github.com/user-attachments/assets/e1d464a7-d473-42db-b24c-31d090cf6d44)

Level1, z = 10m (vegetative in-canopy cells  above surface have Vd> 0)
![image](https://github.com/user-attachments/assets/ebafd821-16a1-4dd9-b5a9-931029052334)

Level2, z = 20m (vegetative in-canopy cells  above surface have Vd> 0 )
![image](https://github.com/user-attachments/assets/46ddb2d4-cea7-437e-8722-1f0fa2bd12dc)

Level3, z = 30m (vegetative in-canopy cells  above surface have Vd> 0)
![image](https://github.com/user-attachments/assets/d896543f-2962-4eaf-9009-34a70b29606e)

z = 40m (vegetative in-canopy cells above surface have Vd> 0; only tallest canopies globally)
![image](https://github.com/user-attachments/assets/ddce3565-d551-418c-856f-5c405dd3d606)

drnimbusrain commented:
@rdsaylor-noaa I have completed all vtypes globally for drydep of all gas species. Full coverage across the globe in terms of drydep gas for all vtypes, including all canopy resolved grid cells (z = 1 to n layers), soils/ground beneath canopies (z=0), and all points outside of contiguous canopies set by NL (i.e.,g barren, snow/ice, urban, and water).  Followed some methods in CMAQ but also some other literature. Results are shown above.  

Thanks again for your continued support in reviewing this PR.  I have one question.  For all vtypes, including vegetative in-canopy (following your ACCESS) and for other vtypes (e.g., barren, snow/ice, urban, and water), I only am including surface resistances (rs) and boundary layer (rb) resistances to get Vd.  However, seems for completeness I am missing **aerodynamic resistances (ra) for Vd = 1/rs + 1/rb + 1/ra**.  Of course we have many ways of including this for vtypes outside of vegetative canopies following MO theory (e.g., even as simple as WSPD/u*), but do you have suggestion on including ra for vegetative canopies, which can deviate from MO theory and certainly are stability dependent?  I thought maybe using sub-canopy diffusivity profile (and its integral) for approximating ra (e.g., Eq. 26 of https://www.sciencedirect.com/science/article/pii/S1352231016307737, following Baldocchi et al.).   Ah, maybe use your ACCESS [aerodynamic resistance from Kv profile](https://github.com/rdsaylor-noaa/ACCESS_FULL/blob/master/src/GlobalData_base.f90#L397), but don't see the formulation immediately in your codes.  I do see an Ra calculation based on stability here:   https://github.com/rdsaylor-noaa/ACCESS_FULL/blob/master/src/EnvironData.f90#L566...so I will probably start with this!  For vertical canopies probably just apply this to the top canopy level only, in some way.

drnimbusrain committed:
Added initial Ra calculation based on ACCESS...not added to Vd yet.

drnimbusrain committed:
Added Ra to drydep calcs, and included snow/ice for urban and water.

drnimbusrain commented:
OK, I have added the[ Ra calculations following ACCESS ](https://github.com/rdsaylor-noaa/ACCESS_FULL/blob/master/src/EnvironData.f90#L566...so), which are now added to the surface resistances and boundary layer resistances for all vtypes, including in-canopy (i.e., Vd = 1/rs + 1/rb + 1/ra).  This Ra calculation is dependent on canopy height, wind speed at reference height above canopy, roughness length and displacement height, and stability (calculated using Bulk Richardson Number, RiB).  

 Some assumptions I made are that for vertically resolved  vegetated canopy grids/points, the ACCESS-based calculated Ra (based on above/near canopy top conditions) is constant throughout the canopy levels down to the ground/soil beneath canopy, and are added as so to all vertical model levels.  For regions outside of vertically resolved canopies, i.e., of other vtypes in canopy-app, to consistently use this Ra calculation I assume a very small "canopy height" close to the ground of 1 cm, which allows for a very small r0 and valid calculations.  Certainly these ideas can be improved later using other approaches (e.g. in-canopy based eddy diffusivity calculation of Ra for vertical canopy columns, q. 26 of https://www.sciencedirect.com/science/article/pii/S1352231016307737, following Baldocchi et al.). Also could update to use a RSL modulated MOL stability in such canopy regions, and map between the RiB and MOL stabilities for consistency.   @rdsaylor-noaa Curious to know your thoughts on these assumptions for now.  
 
 I have also added a boundary for [Ra minimum based on ACCESS by default](https://github.com/rdsaylor-noaa/ACCESS_FULL/blob/d3264e638b593e80a7ce038f957da8326c95e299/src/EnvironData.f90#L239), but it can be changed via the NL.  
 
![image](https://github.com/user-attachments/assets/cf24a7f6-93cc-430e-a024-d16569ddecc9)

drnimbusrain commented:
@rdsaylor-noaa I have completed all initial tasks for this PR and since moved this from draft to ready for review.  I have some preliminary results in slides from today's NOAA canopy project meeting.  See last slides 70-71 for some new animations I just created with latest updates:  https://docs.google.com/presentation/d/1PNtnflK8uZLLDoZ5UcGQE2aQssDcMCitzg835iDQGbY/edit#slide=id.g33b16de7ae6_0_336

rdsaylor commented:
Hi Patrick,

I haven't looked at your code to see exactly what's been done, but my only
caution is that the aerodynamic resistance only applies to the big-leaf
model.  In essence, it accounts for above- and in-canopy transport.  If you
are explicitly calculating transport above and within the canopy, then
there's no need for Ra.  I'm sure you understand this, but I just wanted to
mention it in case.

--
Rick D. Saylor, PhD
Physical Scientist
Atmospheric Turbulence and Diffusion Division
NOAA Air Resources Laboratory
Oak Ridge, TN 37830


On Fri, Mar 21, 2025 at 4:39 PM Patrick Campbell ***@***.***>
wrote:

> OK, I have added the Ra calculations following ACCESS
> <https://github.com/rdsaylor-noaa/ACCESS_FULL/blob/master/src/EnvironData.f90#L566...so>,
> which are now added to the surface resistances and boundary layer
> resistances for all vtypes, including in-canopy. This Ra calculation is
> dependent on canopy height, wind speed at reference height above canopy,
> roughness length and displacement height, and stability (calculated using
> Bulk Richardson Number).
>
> Some assumptions I made are that for vertically resolved vegetated canopy
> grids/points, the ACCESS-based calculated Ra (based on above/near canopy
> top conditions) is constant throughout the canopy levels down to the
> ground/soil beneath canopy, and are added as so to all vertical model
> levels. For regions outside of vertically resolved canopies, i.e., of other
> vtypes in canopy-app, to consistently use this Ra calculation I assume a
> very small "canopy height" close to the ground of 1 cm, which allows for a
> very small r0 and valid calculations. Certainly these ideas can be improved
> later using other approaches (e.g. in-canopy based eddy diffusivity
> calculation of Ra for vertical canopy columns, q. 26 of
> https://www.sciencedirect.com/science/article/pii/S1352231016307737,
> following Baldocchi et al.). @rdsaylor-noaa
> <https://github.com/rdsaylor-noaa> Curious to know your thoughts on these
> assumptions for now.
>
> I have also added a boundary for Ra minimum based on ACCESS by default
> <https://github.com/rdsaylor-noaa/ACCESS_FULL/blob/d3264e638b593e80a7ce038f957da8326c95e299/src/EnvironData.f90#L239>,
> but it can be changed via the NL.
>
> image.png (view on web)
> <https://github.com/user-attachments/assets/cf24a7f6-93cc-430e-a024-d16569ddecc9>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/noaa-oar-arl/canopy-app/pull/149#issuecomment-2744395926>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGCCI52QPKO6C62436VYR732VR2GXAVCNFSM6AAAAABX5RZQE6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBUGM4TKOJSGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: drnimbusrain]*drnimbusrain* left a comment
> (noaa-oar-arl/canopy-app#149)
> <https://github.com/noaa-oar-arl/canopy-app/pull/149#issuecomment-2744395926>
>
> OK, I have added the Ra calculations following ACCESS
> <https://github.com/rdsaylor-noaa/ACCESS_FULL/blob/master/src/EnvironData.f90#L566...so>,
> which are now added to the surface resistances and boundary layer
> resistances for all vtypes, including in-canopy. This Ra calculation is
> dependent on canopy height, wind speed at reference height above canopy,
> roughness length and displacement height, and stability (calculated using
> Bulk Richardson Number).
>
> Some assumptions I made are that for vertically resolved vegetated canopy
> grids/points, the ACCESS-based calculated Ra (based on above/near canopy
> top conditions) is constant throughout the canopy levels down to the
> ground/soil beneath canopy, and are added as so to all vertical model
> levels. For regions outside of vertically resolved canopies, i.e., of other
> vtypes in canopy-app, to consistently use this Ra calculation I assume a
> very small "canopy height" close to the ground of 1 cm, which allows for a
> very small r0 and valid calculations. Certainly these ideas can be improved
> later using other approaches (e.g. in-canopy based eddy diffusivity
> calculation of Ra for vertical canopy columns, q. 26 of
> https://www.sciencedirect.com/science/article/pii/S1352231016307737,
> following Baldocchi et al.). @rdsaylor-noaa
> <https://github.com/rdsaylor-noaa> Curious to know your thoughts on these
> assumptions for now.
>
> I have also added a boundary for Ra minimum based on ACCESS by default
> <https://github.com/rdsaylor-noaa/ACCESS_FULL/blob/d3264e638b593e80a7ce038f957da8326c95e299/src/EnvironData.f90#L239>,
> but it can be changed via the NL.
>
> image.png (view on web)
> <https://github.com/user-attachments/assets/cf24a7f6-93cc-430e-a024-d16569ddecc9>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/noaa-oar-arl/canopy-app/pull/149#issuecomment-2744395926>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGCCI52QPKO6C62436VYR732VR2GXAVCNFSM6AAAAABX5RZQE6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBUGM4TKOJSGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

drnimbusrain commented:
> Hi Patrick, I haven't looked at your code to see exactly what's been done, but my only caution is that the aerodynamic resistance only applies to the big-leaf model. In essence, it accounts for above- and in-canopy transport. If you are explicitly calculating transport above and within the canopy, then there's no need for Ra. I'm sure you understand this, but I just wanted to mention it in case.
> […](#)
> -- Rick D. Saylor, PhD Physical Scientist Atmospheric Turbulence and Diffusion Division NOAA Air Resources Laboratory Oak Ridge, TN 37830
> On Fri, Mar 21, 2025 at 4:39 PM Patrick Campbell ***@***.***> wrote: OK, I have added the Ra calculations following ACCESS <https://github.com/rdsaylor-noaa/ACCESS_FULL/blob/master/src/EnvironData.f90#L566...so>, which are now added to the surface resistances and boundary layer resistances for all vtypes, including in-canopy. This Ra calculation is dependent on canopy height, wind speed at reference height above canopy, roughness length and displacement height, and stability (calculated using Bulk Richardson Number). Some assumptions I made are that for vertically resolved vegetated canopy grids/points, the ACCESS-based calculated Ra (based on above/near canopy top conditions) is constant throughout the canopy levels down to the ground/soil beneath canopy, and are added as so to all vertical model levels. For regions outside of vertically resolved canopies, i.e., of other vtypes in canopy-app, to consistently use this Ra calculation I assume a very small "canopy height" close to the ground of 1 cm, which allows for a very small r0 and valid calculations. Certainly these ideas can be improved later using other approaches (e.g. in-canopy based eddy diffusivity calculation of Ra for vertical canopy columns, q. 26 of https://www.sciencedirect.com/science/article/pii/S1352231016307737, following Baldocchi et al.). @rdsaylor-noaa <https://github.com/rdsaylor-noaa> Curious to know your thoughts on these assumptions for now. I have also added a boundary for Ra minimum based on ACCESS by default <https://github.com/rdsaylor-noaa/ACCESS_FULL/blob/d3264e638b593e80a7ce038f957da8326c95e299/src/EnvironData.f90#L239>, but it can be changed via the NL. image.png (view on web) <https://github.com/user-attachments/assets/cf24a7f6-93cc-430e-a024-d16569ddecc9> — Reply to this email directly, view it on GitHub <[#149 (comment)](https://github.com/noaa-oar-arl/canopy-app/pull/149#issuecomment-2744395926)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGCCI52QPKO6C62436VYR732VR2GXAVCNFSM6AAAAABX5RZQE6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBUGM4TKOJSGY> . You are receiving this because you were mentioned.Message ID: ***@***.***> [image: drnimbusrain]*drnimbusrain* left a comment ([noaa-oar-arl/canopy-app#149](https://github.com/noaa-oar-arl/canopy-app/pull/149)) <[#149 (comment)](https://github.com/noaa-oar-arl/canopy-app/pull/149#issuecomment-2744395926)> OK, I have added the Ra calculations following ACCESS <https://github.com/rdsaylor-noaa/ACCESS_FULL/blob/master/src/EnvironData.f90#L566...so>, which are now added to the surface resistances and boundary layer resistances for all vtypes, including in-canopy. This Ra calculation is dependent on canopy height, wind speed at reference height above canopy, roughness length and displacement height, and stability (calculated using Bulk Richardson Number). Some assumptions I made are that for vertically resolved vegetated canopy grids/points, the ACCESS-based calculated Ra (based on above/near canopy top conditions) is constant throughout the canopy levels down to the ground/soil beneath canopy, and are added as so to all vertical model levels. For regions outside of vertically resolved canopies, i.e., of other vtypes in canopy-app, to consistently use this Ra calculation I assume a very small "canopy height" close to the ground of 1 cm, which allows for a very small r0 and valid calculations. Certainly these ideas can be improved later using other approaches (e.g. in-canopy based eddy diffusivity calculation of Ra for vertical canopy columns, q. 26 of https://www.sciencedirect.com/science/article/pii/S1352231016307737, following Baldocchi et al.). @rdsaylor-noaa <https://github.com/rdsaylor-noaa> Curious to know your thoughts on these assumptions for now. I have also added a boundary for Ra minimum based on ACCESS by default <https://github.com/rdsaylor-noaa/ACCESS_FULL/blob/d3264e638b593e80a7ce038f957da8326c95e299/src/EnvironData.f90#L239>, but it can be changed via the NL. image.png (view on web) <https://github.com/user-attachments/assets/cf24a7f6-93cc-430e-a024-d16569ddecc9> — Reply to this email directly, view it on GitHub <[#149 (comment)](https://github.com/noaa-oar-arl/canopy-app/pull/149#issuecomment-2744395926)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGCCI52QPKO6C62436VYR732VR2GXAVCNFSM6AAAAABX5RZQE6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBUGM4TKOJSGY> . You are receiving this because you were mentioned.Message ID: ***@***.***>

@rdsaylor-noaa Thank you for your valuable insight here.  Yes, aware of it, but canopy-app does not link calculated transport above/within canopy to dry dep here.  We have the option to calculate the eddy K profile above/within canopy (```ifcaneddy```), but it is not used in dry deposition option for explicit transport of species.

So, right now I calculate Ra  (based on driving GFS model winds etc.) and add it similarly to the resistances for all levels at canopy top and within canopy (and at ground).  Please let me know what you think further.  Otherwise I can merge in and we can fix later.

drnimbusrain commented:
@rdsaylor-noaa I am going to merge this into develop to move it forward, but we can improve later with Ra and other approximations and enhancements.
